python/task_manager/worker_thread.py

```diff
@@ -43,24 +44,18 @@ def run_task(self, task):
 
         :param task:    The task to run
         """
-        self._mutex.lock()
-        try:
+        with self._mutex:
             self._task = task
-        finally:
-            self._mutex.unlock()
-        self._wait_condition.wakeAll()
+            self._wait_condition.notifyAll()
 
     def shut_down(self):
         """
         Shut down the thread and wait for it to exit before returning
         """
         self._results_dispatcher = None
-        self._mutex.lock()
-        try:
+        with self._mutex:
             self._process_tasks = False
-        finally:
-            self._mutex.unlock()
-        self._wait_condition.wakeAll()
+            self._wait_condition.notifyAll()
```

Comment:
With `threading.Condition`, you need to own the lock to be able to notify, which if fine here, as the lock is released when notifying so someone else can grab it.